service/collector/nat.go

```diff
@@ -0,0 +1,270 @@
+package collector
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/hprose/hprose-go"
+	"github.com/prometheus/client_golang/prometheus"
+
+	clientaws "github.com/giantswarm/aws-operator/client/aws"
+	"github.com/giantswarm/aws-operator/service/controller/key"
+	"github.com/giantswarm/aws-operator/service/internal/cache"
+)
+
+const (
+	// __NATCache__ is used as temporal cache key to save NAT response.
+	prefixNATcacheKey = "__NATCache__"
+	labelVPC          = "vpc"
+	labelAZ           = "availability_zone"
+)
+
+const (
+	subsystemNAT = "nat"
+)
+
+var (
+	natDesc *prometheus.Desc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, subsystemNAT, "info"),
+		"NAT limits information.",
+		[]string{
+			labelAccountID,
+			labelVPC,
+			labelAZ,
+		},
+		nil,
+	)
+)
+
+type NATConfig struct {
+	Helper *helper
+	Logger micrologger.Logger
+
+	InstallationName string
+}
+
+type NAT struct {
+	cache  *natCache
+	helper *helper
+	logger micrologger.Logger
+
+	installationName string
+}
+
+type natCache struct {
+	cache *cache.StringCache
+}
+
+type natInfoResponse struct {
+	vpcs map[string]vpcInfo
+}
+
+type vpcInfo struct {
+	natGatewaysByZone map[string]float64
+}
+
+func NewNAT(config NATConfig) (*NAT, error) {
+	if config.Helper == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Helper must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.InstallationName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.InstallationName must not be empty", config)
+	}
+
+	v := &NAT{
+		// AWS operator creates at this moment one NAT for each private subnet (node
+		// pool). As clusters are not created nor changed so often, and the process
+		// can take around 20 minutes, 30 minutes for the cache expiration is a
+		// reasonable value.
+		cache:  newNATCache(time.Minute * 30),
+		helper: config.Helper,
+		logger: config.Logger,
+
+		installationName: config.InstallationName,
+	}
+
+	return v, nil
+}
+
+func newNATCache(expiration time.Duration) *natCache {
+	cache := &natCache{
+		cache: cache.NewStringCache(expiration),
+	}
+
+	return cache
+}
+
+func (n *natCache) Get(accID string) (*natInfoResponse, error) {
+	var c natInfoResponse
+	raw, ok := n.cache.Get(prefixNATcacheKey + accID)
+	if ok {
+		err := hprose.Unserialize(raw, c, true)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	return &c, nil
+}
+
+func (n *natCache) Set(accID string, content natInfoResponse) error {
+	contentSerialized, err := hprose.Serialize(content, true)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	n.cache.Set(prefixNATcacheKey+accID, contentSerialized)
```

Comment:
The content string is saved with the common prefix and the account ID (each account store different API response)

Reply:
Minor note, but it might improve maintainability if you extract a function for generating the cache key from `accID`. That way you only need to change one place instead of finding different places where the cache is utilized.